plugins/inputs/ds389/ds389.go

```diff
@@ -0,0 +1,373 @@
+			acc.AddError(fmt.Errorf("Invalid setting for ssl: %s", o.TLS))
+			return nil
+		}
+	} else {
+		l, err = ldap.Dial("tcp", fmt.Sprintf("%s:%d", o.Host, o.Port))
+	}
+
+	if err != nil {
+		acc.AddError(err)
+		return nil
+	}
+	defer l.Close()
+
+	// username/password bind
+	if o.BindDn != "" && o.BindPassword != "" {
+		err = l.Bind(o.BindDn, o.BindPassword)
+		if err != nil {
+			acc.AddError(err)
+			return nil
+		}
+	}
+
+	searchRequest := ldap.NewSearchRequest(
+		searchMonitor,
+		ldap.ScopeWholeSubtree,
+		ldap.NeverDerefAliases,
+		0,
+		0,
+		false,
+		searchFilter,
+		searchAttrs,
+		nil,
+	)
+
+	sr, err := l.Search(searchRequest)
+
+	if err != nil {
+		acc.AddError(err)
+		return nil
+	}
+
+	version := sr.Entries[0].GetAttributeValue("version")
+	field := gatherSearchResult(sr, o.Status)
+
+	searchLdbmRequest := ldap.NewSearchRequest(
+		searchLdbmMonitor,
+		ldap.ScopeWholeSubtree,
+		ldap.NeverDerefAliases,
+		0,
+		0,
+		false,
+		searchFilter,
+		searchLdbmAttrs,
+		nil,
+	)
+
+	sldbmr, err := l.Search(searchLdbmRequest)
+
+	if err != nil {
+		acc.AddError(err)
+		return nil
+	}
+
+	for k, v := range gatherSearchResult(sldbmr, false) {
+		field[k] = v
+	}
+
+	if o.AllDbmonitor {
+		for _, searchDbMonitor := range sr.Entries[0].GetAttributeValues("backendmonitordn") {
+			searchDbRequest := ldap.NewSearchRequest(
+				searchDbMonitor,
+				ldap.ScopeWholeSubtree,
+				ldap.NeverDerefAliases,
+				0,
+				0,
+				false,
+				searchFilter,
+				searchDbAttrs,
+				nil,
+			)
+
+			sdbr, err := l.Search(searchDbRequest)
+			if err != nil {
+				acc.AddError(err)
+				return nil
+			}
+			r := regexp.MustCompile(`cn=monitor,cn=(?P<db>\w+),cn=ldbm database,cn=plugins,cn=config`)
+			db := r.FindStringSubmatch(searchDbMonitor)[1]
+			for k, v := range gatherDbSearchResult(sdbr, db) {
+				field[k] = v
+			}
+		}
+	} else {
+		if len(o.Dbtomonitor) > 0 {
+			for _, db := range o.Dbtomonitor {
+				var searchDbMonitor = fmt.Sprintf("cn=monitor,cn=%s,cn=ldbm database,cn=plugins,cn=config", db)
+				searchDbRequest := ldap.NewSearchRequest(
+					searchDbMonitor,
+					ldap.ScopeWholeSubtree,
+					ldap.NeverDerefAliases,
+					0,
+					0,
+					false,
+					searchFilter,
+					searchDbAttrs,
+					nil,
+				)
+
+				sdbr, err := l.Search(searchDbRequest)
+				if err != nil {
+					acc.AddError(err)
+					return nil
+				}
+				for k, v := range gatherDbSearchResult(sdbr, db) {
+					field[k] = v
+				}
+			}
+		}
+	}
+
+	// Add metrics
+	tags := map[string]string{
+		"server":  o.Host,
+		"port":    strconv.Itoa(o.Port),
+		"version": version,
+	}
+	acc.AddFields("ds389", field, tags)
+	return nil
+}
+
+func gatherSearchResult(sr *ldap.SearchResult, status bool) map[string]interface{} {
+	fields := map[string]interface{}{}
+	for _, entry := range sr.Entries {
+		for _, attr := range entry.Attributes {
+			if attr.Name == "connection" && status {
+				for _, thisAttr := range attr.Values {
+					elements := strings.Split(thisAttr, ":")
```

Comment:
Verify the length of `elements` to avoid a potential panic when indexing the slice.